Add base class for single-subgraph inductor HOPs

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #139928
* __->__ #139898

This PR adds "PrimHOPBase", which is intended to be a base class that
one can extend to create new HOPs that match some criteria:
- they take one subgraph as input, and their semantics are running the
  subgraph on some operands
- the HOP stays alive until Inductor

The motivation is that we are seeing a lot more HOPs (invoke_subgraph,
invoke_quant) that have this property and there can be a lot of shared
code between them.

Future:
- Migrate invoke_subgraph to use this
- There are some TODOs in the code

Test Plan:
- new tests

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

zou3519 committed:
Add base class for single-subgraph inductor HOPs

This PR adds "PrimHOPBase", which is intended to be a base class that
one can extend to create new HOPs that match some criteria:
- they take one subgraph as input, and their semantics are running the
  subgraph on some operands
- the HOP stays alive until Inductor

The motivation is that we are seeing a lot more HOPs (invoke_subgraph,
invoke_quant) that have this property and there can be a lot of shared
code between them.

Future:
- Migrate invoke_subgraph to use this
- There are some TODOs in the code

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add base class for single-subgraph inductor HOPs"


This PR adds "PrimHOPBase", which is intended to be a base class that
one can extend to create new HOPs that match some criteria:
- they take one subgraph as input, and their semantics are running the
  subgraph on some operands
- the HOP stays alive until Inductor

The motivation is that we are seeing a lot more HOPs (invoke_subgraph,
invoke_quant) that have this property and there can be a lot of shared
code between them.

Future:
- Migrate invoke_subgraph to use this
- There are some TODOs in the code

Test Plan:
- new tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

zou3519 committed:
Update on "Add base class for single-subgraph inductor HOPs"


This PR adds "PrimHOPBase", which is intended to be a base class that
one can extend to create new HOPs that match some criteria:
- they take one subgraph as input, and their semantics are running the
  subgraph on some operands
- the HOP stays alive until Inductor

The motivation is that we are seeing a lot more HOPs (invoke_subgraph,
invoke_quant) that have this property and there can be a lot of shared
code between them.

Future:
- Migrate invoke_subgraph to use this
- There are some TODOs in the code

Test Plan:
- new tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

zou3519 committed:
Update on "Add base class for single-subgraph inductor HOPs"


This PR adds "PrimHOPBase", which is intended to be a base class that
one can extend to create new HOPs that match some criteria:
- they take one subgraph as input, and their semantics are running the
  subgraph on some operands
- the HOP stays alive until Inductor

The motivation is that we are seeing a lot more HOPs (invoke_subgraph,
invoke_quant) that have this property and there can be a lot of shared
code between them.

Future:
- Migrate invoke_subgraph to use this
- There are some TODOs in the code

Test Plan:
- new tests

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

zou3519 committed:
Update on "Add base class for single-subgraph inductor HOPs"


This PR adds "PrimHOPBase", which is intended to be a base class that
one can extend to create new HOPs that match some criteria:
- they take one subgraph as input, and their semantics are running the
  subgraph on some operands
- the HOP stays alive until Inductor

The motivation is that we are seeing a lot more HOPs (invoke_subgraph,
invoke_quant) that have this property and there can be a lot of shared
code between them.

Future:
- Migrate invoke_subgraph to use this
- There are some TODOs in the code

Test Plan:
- new tests

[ghstack-poisoned]

anijain2305 commented on torch/_higher_order_ops/prim_hop_base.py:
```diff
@@ -0,0 +1,178 @@
+        # TODO: Something special needs to happen with min cut partitioner
+        with suspend_functionalization(), disable_functional_mode(), torch.enable_grad():
+            with disable_proxy_modes_tracing():
+                from .utils import _from_fun, create_fw_bw_graph
```
I would like this to be invoke_subgraph `create_fw_bw_graph` (or replace utils.py one with the invoke_subgraph one). The main reason is the signatures for the joint graph and ergonomics that come with signatures.

For the joint graph from invoke_subgraph
1) Input signature is -  (*primals, *tangents).
2) Output signature is - (*grads, *fw_outs)

The nice thing about input signature is that it is already in a form that min-cut partitioner expects.

The nice thing about output signature is that `grads` are already at the right place. Once we apply min-cut partitioner, the new output signature of the backward graph would be just (*grads). So, the graph transformation is easy. All the getitems on the output of the old joint_graph are still valid for the new_bw graph. This reduces LOC and also chances of error.

zou3519 replied:
Yeah I can swap this out

zou3519 replied:
Updated to use invoke_subgraph's create_fw_bw_graph

zou3519 committed:
Update on "Add base class for single-subgraph inductor HOPs"


This PR adds "PrimHOPBase", which is intended to be a base class that
one can extend to create new HOPs that match some criteria:
- they take one subgraph as input, and their semantics are running the
  subgraph on some operands
- the HOP stays alive until Inductor

The motivation is that we are seeing a lot more HOPs (invoke_subgraph,
invoke_quant) that have this property and there can be a lot of shared
code between them.

Future:
- Migrate invoke_subgraph to use this
- There are some TODOs in the code

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge

zou3519 commented on torch/_higher_order_ops/prim_hop_base.py:
```diff
@@ -0,0 +1,190 @@
+
+                fw_inputs = pytree.tree_map(_from_fun, operands)
+                fw_outputs = subgraph(*fw_inputs)
+                _, joint_graph, _ = create_fw_bw_graph(
```
It's more of a long-term plan that we only materialize in proxy tensor key. I'm not sure how much it'll take to get there.
